### PR TITLE
[3.10] Joomla.Showon.initialise API, and  make sure that showon initialised only once, fork of 18843

### PR DESCRIPTION
**Pull request by Fedik**
Pull Request for Issue #18843 and only for J3 .

### Summary of Changes

This is fork of #18843, with requested changes.
The code make sure that showon initialised only once, no mater how often Joomla.setUpShowon was called.
This is achieved with extra flag `showonInitialised` that sets while showon initialisation.

### Testing Instructions

Make sure `showon` still works.

Example, go to global config change "Debug Language" on/off, should downup/hides "Language Display" field.
Also the field "Language Display"  will have extra attribute `data-showon-initialised`

### Actual result BEFORE applying this Pull Request
showon works,
the showon field do not have `data-showon-initialised`

### Expected result AFTER applying this Pull Request
showon works,
the showon field do have `data-showon-initialised` attribute

### CODE REVIEW

**Comment by richard67** [comment]
@Fedik #18843 can be closed, right?

**Comment by Fedik** [comment]
yeah, I have included changes from that pull

**Comment by dgrammatiko** [comment]
@Fedik diffing the changes with the v4 version of show on I think you need to rename this to `Joomla.Showon.initilise` to be called like `Joomla.Showon.initilise(field)`. The initialise in J4 could be a function in the Class with this code (the refactoring should be straight forward):
https://github.com/joomla/joomla-cms/blob/d49e4614f299b5f158f6a585de6f9c0313013168/build/media_source/system/js/showon.es6.js#L32-L84

**Comment by Fedik** [comment]
This only for J3, it should not go to J4,
there already event stuff for such thing:
https://github.com/joomla/joomla-cms/blob/d49e4614f299b5f158f6a585de6f9c0313013168/build/media_source/system/js/showon.es6.js#L215-L218

Only missed check for "double initialisation", I can do it later in another pull.

I do not think that this patch is much needed, but people asked :wink: 
It can be safely closed, depend what @HLeithner will decide

**Comment by HLeithner** [comment]
@Fedik would be cool if we have the double initialization added in this pr too. After 2 tests we can merge it in to 3.10 (I think that makes more sense). I don't expect another 3.9 release (hopefully). It also introduces a new API endpoint so 3.10 is a better target.

**Comment by Fedik** [comment]
> would be cool if we have the double initialization added in this pr too

it is here, that why I made this fork:
https://github.com/joomla/joomla-cms/blob/0f5956e99c8f859e0668b790a5e31629b964688b/media/jui/js/cms-uncompressed.js#L184-L189

I meant that for J4 will be need a separted PR.

> It also introduces a new API endpoint so 3.10 is a better target.

I can remove Joomla.setUpShowon and leave "event" $(document).on('subform-row-add'), I guess most people who need it already use in this way, so it will be "bug fix" :wink:

File: media/jui/js/cms-uncompressed.js [hunk 1]
```diff
@@ -209,31 +209,47 @@ Joomla = window.Joomla || {};
...
-
-				// Fix showon field names in a current group
-				for (var i = 0, l = $elements.length; i < l; i++) {
-					$elm   = $($elements[i]);
-					showon = $elm.attr('data-showon').replace(search, replace);
-
-					$elm.attr('data-showon', showon);
+			Joomla.Showon.initilise(document);
+
+            // Setup showon feature in the modified container
```
Comment (richard67) [inline media/jui/js/cms-uncompressed.js]:
```suggestion
			// Setup showon feature in the modified container
```

**Comment by Fedik** [comment]
Okay, so, I have backport Joomla 4 'joomla:updated' stuff for showon, now it can be more easy for transition.
And added `Joomla.Showon.initilise` as @dgrammatiko suggested, later I will make pull  for J4.

Testing the same, make sure `showon` still work.

Please someone change target branch to joomla-10, or I have to make new PR?

**Comment by HLeithner** [comment]
I change the branch for you.

But you can do this your self by adding the PR
image

Thanks for your PR, now we need some tests ;-)

**Comment by Fedik** [comment]
~well, I just broke PR~ 
~need some time to fix~

okay, I think I need to sync with j 3.10 branch

**Comment by Fedik** [comment]
@HLeithner can I merge joomla:3.10-dev to my branch for sync? or I need make new

**Comment by HLeithner** [comment]
I updated your branch you only need to pull from github again.

**Comment by HLeithner** [comment]
but yes you can simple merge 3.10-dev into your branch

**Comment by Fedik** [comment]
okay, thanks

**Comment by richard67** [comment]
@HLeithner Now the PR shows more changes than only those from @Fedik . It shows the changes which meanwhile have been made in the staging branch since the last upmerge of staging into 3-10-dev. So maybe it's easier just to redo the PR for 3.-10-dev?

**Comment by Fedik** [comment]
@richard67 or we can wait when `staging` will be merged in to 3-10-dev :)

**Comment by richard67** [comment]
Yes, that will also work.

**Comment by HLeithner** [comment]
I will upmerge staging

**Comment by HLeithner** [comment]
so now it's ok again

**Comment by richard67** [comment]
Yes, looks ok now.

**Comment by Fedik** [comment]
I have updated testing instruction, and made same changes for joomla 4 #32069

File: media/jui/js/cms-uncompressed.js [hunk 2]
```diff
@@ -195,35 +204,52 @@ Joomla = window.Joomla || {};
...
-				})();
+
+				})($($showonFields[is]));
+			}
+		}
+
+		// Provide a public API
+		if (!Joomla.Showon) {
+			Joomla.Showon = {
+				initilise: setUpShowon
```
Comment (dgrammatiko) [inline media/jui/js/cms-uncompressed.js]: Ooops typo: `initialise `

Comment (Fedik) [reply media/jui/js/cms-uncompressed.js]: fixed

**Comment by dgrammatiko** [comment]
I have tested this item :white_check_mark: successfully on 04e2d9de9fe18daca3a3cb70f4a66ab864505129<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32061">issues.joomla.org/tracker/joomla-cms/32061</a>.</sub>

**Comment by ghost** [comment]
I have tested this item :white_check_mark: successfully on 04e2d9de9fe18daca3a3cb70f4a66ab864505129<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32061">issues.joomla.org/tracker/joomla-cms/32061</a>.</sub>

**Comment by richard67** [comment]
RTC<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/32061">issues.joomla.org/tracker/joomla-cms/32061</a>.</sub>

**Comment by zero-24** [comment]
@Fedik do we need to add documentation for this change somewhere? And if so what should be documented?

**Comment by Fedik** [comment]
@zero-24 good question, I have no idea ;) 
Do we have any documentation about `showon` API somwhere? if no, then maybe we need to add.

Technically and partly it is a new feature, to cover edge cases from #18843 , also see comment  https://github.com/joomla/joomla-cms/pull/18843#issuecomment-346930106

In J4 we already have events #16016 (I backported it partly here), that handle "dynamic" content changes

**Comment by zero-24** [comment]
Ok will merge than this heree for 3.10. thanks

**Pull request merged**